internal/app/export/service.go

```diff
@@ -4,62 +4,294 @@
 package export
 
 import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
 	"fmt"
+	"log"
+	"strings"
+	"sync"
+	"sync/atomic"
 
 	mqtt "github.com/eclipse/paho.mqtt.golang"
+	"github.com/mainflux/export/internal/pkg/messages"
 	"github.com/mainflux/export/internal/pkg/routes"
 	"github.com/mainflux/export/internal/pkg/routes/mfx"
 	"github.com/mainflux/export/pkg/config"
-	log "github.com/mainflux/mainflux/logger"
+	logger "github.com/mainflux/mainflux/logger"
 	nats "github.com/nats-io/nats.go"
 )
 
 type Service interface {
-	Start(queue string, nc *nats.Conn)
+	Start(queue string) error
+	Subscribe(topic string, nc *nats.Conn)
 }
 
 var _ Service = (*exporter)(nil)
 
 type exporter struct {
-	Mqtt   mqtt.Client
-	Logger log.Logger
-	Cfg    config.Config
-	Routes []routes.Route
+	ID        string
+	MQTT      mqtt.Client
+	Logger    logger.Logger
+	Cfg       config.Config
+	Consumers map[string]routes.Route
+	Cache     messages.Cache
+	connected chan bool
+	status    uint32
+	sync.RWMutex
 }
 
+const (
+	exportGroup = "export-group"
+	count       = 100
+
+	disconnected uint32 = iota
+	connected
+
+	NatsSub = "export"
+	NatsAll = ">"
+)
+
+var errNoRoutesConfigured = errors.New("No routes configured")
+
 // New create new instance of export service
-func New(mqtt mqtt.Client, c config.Config, logger log.Logger) Service {
-	routes := make([]routes.Route, 0)
+func New(c config.Config, cache messages.Cache, l logger.Logger) Service {
+	routes := make(map[string]routes.Route)
+	id := fmt.Sprintf("export-%s", c.MQTT.Username)
+
 	e := exporter{
-		Mqtt:   mqtt,
-		Logger: logger,
-		Cfg:    c,
-		Routes: routes,
+		ID:        id,
+		Logger:    l,
+		Cfg:       c,
+		Consumers: routes,
+		Cache:     cache,
+		connected: make(chan bool, 1),
 	}
-
+	client, err := e.mqttConnect(c, l)
+	if err != nil {
+		log.Fatalf(err.Error())
+	}
+	e.MQTT = client
 	return &e
 }
 
-// Start method starts consuming messages received from NATS
-// and makes routes according to the configuration file.
-// Routes export messages to mqtt.
-func (e *exporter) Start(queue string, nc *nats.Conn) {
+// Start method loads route configuration
+func (e *exporter) Start(queue string) error {
 	var route routes.Route
 	for _, r := range e.Cfg.Routes {
+		natsTopic := fmt.Sprintf("%s.%s", NatsSub, r.NatsTopic)
 		switch r.Type {
 		case "mfx":
```

Comment:
Constants (iota)? If there are only two cases (one of which is the default), there is no need for this and you can extract it in the future if some other types came up.

Reply:
dont quite get this, this is route from config.toml

```
[[routes]]
  mqtt_topic = "channels/2a362536-e835-44de-be73-e5cf3df1a0f1/messages"
  nats_topic = "test.test"
  subtopic = ""
  type = "plain"
```

Reply:
I was talking about this switch-case (`switch r.Type`). If you think that you don't need consts, you can ignore this comment.

Reply:
ok, I think it is ok for now like this
we will extend this later with additinal types and this part will evolve